src/EditorFeatures/Core/Implementation/CodeDefinitionWindow/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,238 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.SymbolMapping;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.CodeDefinitionWindow
+{
+    [Export(typeof(ITextViewConnectionListener))]
+    [Export(typeof(DefinitionContextTracker))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
+    internal class DefinitionContextTracker : ITextViewConnectionListener
+    {
+        private readonly HashSet<ITextView> _subscribedViews = new HashSet<ITextView>();
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+        private readonly ICodeDefinitionWindowService _codeDefinitionWindowService;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsynchronousOperationListener _asyncListener;
+
+        private CancellationTokenSource? _currentUpdateCancellationToken;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DefinitionContextTracker(
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            ICodeDefinitionWindowService codeDefinitionWindowService,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+            _threadingContext = threadingContext;
+            _asyncListener = listenerProvider.GetListener(FeatureAttribute.CodeDefinitionWindow);
+        }
+
+        void ITextViewConnectionListener.SubjectBuffersConnected(ITextView textView, ConnectionReason reason, IReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            Contract.ThrowIfFalse(_threadingContext.JoinableTaskContext.IsOnMainThread);
+
+            if (!_subscribedViews.Contains(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _subscribedViews.Add(textView);
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
```

Comment:
```suggestion
            if (!textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView) &&
                _subscribedViews.Add(textView))
            {
                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
```

up to youif that's better.  i can see arguments that it's less clear.

Reply:
Oh yeah ick, let's not do that. :smile: